// @juicyllama/utils/dist/enums/currencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupportedCurrencies = void 0;
var SupportedCurrencies;
(function (SupportedCurrencies) {
    SupportedCurrencies["AUD"] = "AUD";
    SupportedCurrencies["CAD"] = "CAD";
    SupportedCurrencies["CHF"] = "CHF";
    SupportedCurrencies["EUR"] = "EUR";
    SupportedCurrencies["GBP"] = "GBP";
    SupportedCurrencies["INR"] = "INR";
    SupportedCurrencies["MXN"] = "MXN";
    SupportedCurrencies["USD"] = "USD";
    SupportedCurrencies["JPY"] = "JPY";
    SupportedCurrencies["CNY"] = "CNY";
    SupportedCurrencies["HKD"] = "HKD";
    SupportedCurrencies["NZD"] = "NZD";
    SupportedCurrencies["SEK"] = "SEK";
    SupportedCurrencies["KRW"] = "KRW";
    SupportedCurrencies["SGD"] = "SGD";
    SupportedCurrencies["NOK"] = "NOK";
    SupportedCurrencies["RUB"] = "RUB";
    SupportedCurrencies["ZAR"] = "ZAR";
    SupportedCurrencies["TRY"] = "TRY";
    SupportedCurrencies["BRL"] = "BRL";
})(SupportedCurrencies || (exports.SupportedCurrencies = SupportedCurrencies = {}));
