// @juicyllama/utils/dist/enums/dates.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MonthNamesShort = exports.MonthNames = exports.DayNames = exports.FrequencyPerYear = exports.SubscriptionFrequency = void 0;
var SubscriptionFrequency;
(function (SubscriptionFrequency) {
    SubscriptionFrequency[SubscriptionFrequency["DAILY"] = 1] = "DAILY";
    SubscriptionFrequency[SubscriptionFrequency["WEEKLY"] = 7] = "WEEKLY";
    SubscriptionFrequency[SubscriptionFrequency["BIWEEKLY"] = 14] = "BIWEEKLY";
    SubscriptionFrequency[SubscriptionFrequency["MONTHLY"] = 30] = "MONTHLY";
    SubscriptionFrequency[SubscriptionFrequency["BIMONTHLY"] = 60] = "BIMONTHLY";
    SubscriptionFrequency[SubscriptionFrequency["QUARTERLY"] = 90] = "QUARTERLY";
    SubscriptionFrequency[SubscriptionFrequency["BIQUARTERLY"] = 180] = "BIQUARTERLY";
    SubscriptionFrequency[SubscriptionFrequency["YEARLY"] = 365] = "YEARLY";
})(SubscriptionFrequency || (exports.SubscriptionFrequency = SubscriptionFrequency = {}));
var FrequencyPerYear;
(function (FrequencyPerYear) {
    FrequencyPerYear[FrequencyPerYear["DAILY"] = 365] = "DAILY";
    FrequencyPerYear[FrequencyPerYear["TWICE_PER_WEEK"] = 104] = "TWICE_PER_WEEK";
    FrequencyPerYear[FrequencyPerYear["WEEKLY"] = 52] = "WEEKLY";
    FrequencyPerYear[FrequencyPerYear["TWICE_PER_MONTH"] = 24] = "TWICE_PER_MONTH";
    FrequencyPerYear[FrequencyPerYear["MONTHLY"] = 12] = "MONTHLY";
    FrequencyPerYear[FrequencyPerYear["QUARTERLY"] = 4] = "QUARTERLY";
    FrequencyPerYear[FrequencyPerYear["YEARLY"] = 1] = "YEARLY";
    FrequencyPerYear[FrequencyPerYear["NONE"] = 0] = "NONE";
})(FrequencyPerYear || (exports.FrequencyPerYear = FrequencyPerYear = {}));
var DayNames;
(function (DayNames) {
    DayNames[DayNames["Monday"] = 0] = "Monday";
    DayNames[DayNames["Tuesday"] = 1] = "Tuesday";
    DayNames[DayNames["Wednesday"] = 2] = "Wednesday";
    DayNames[DayNames["Thursday"] = 3] = "Thursday";
    DayNames[DayNames["Friday"] = 4] = "Friday";
    DayNames[DayNames["Saturday"] = 5] = "Saturday";
    DayNames[DayNames["Sunday"] = 6] = "Sunday";
})(DayNames || (exports.DayNames = DayNames = {}));
var MonthNames;
(function (MonthNames) {
    MonthNames[MonthNames["January"] = 0] = "January";
    MonthNames[MonthNames["February"] = 1] = "February";
    MonthNames[MonthNames["March"] = 2] = "March";
    MonthNames[MonthNames["April"] = 3] = "April";
    MonthNames[MonthNames["May"] = 4] = "May";
    MonthNames[MonthNames["June"] = 5] = "June";
    MonthNames[MonthNames["July"] = 6] = "July";
    MonthNames[MonthNames["August"] = 7] = "August";
    MonthNames[MonthNames["September"] = 8] = "September";
    MonthNames[MonthNames["October"] = 9] = "October";
    MonthNames[MonthNames["November"] = 10] = "November";
    MonthNames[MonthNames["December"] = 11] = "December";
})(MonthNames || (exports.MonthNames = MonthNames = {}));
var MonthNamesShort;
(function (MonthNamesShort) {
    MonthNamesShort[MonthNamesShort["Jan"] = 0] = "Jan";
    MonthNamesShort[MonthNamesShort["Feb"] = 1] = "Feb";
    MonthNamesShort[MonthNamesShort["Mar"] = 2] = "Mar";
    MonthNamesShort[MonthNamesShort["Apr"] = 3] = "Apr";
    MonthNamesShort[MonthNamesShort["May"] = 4] = "May";
    MonthNamesShort[MonthNamesShort["Jun"] = 5] = "Jun";
    MonthNamesShort[MonthNamesShort["Jul"] = 6] = "Jul";
    MonthNamesShort[MonthNamesShort["Aug"] = 7] = "Aug";
    MonthNamesShort[MonthNamesShort["Sep"] = 8] = "Sep";
    MonthNamesShort[MonthNamesShort["Oct"] = 9] = "Oct";
    MonthNamesShort[MonthNamesShort["Nov"] = 10] = "Nov";
    MonthNamesShort[MonthNamesShort["Dec"] = 11] = "Dec";
})(MonthNamesShort || (exports.MonthNamesShort = MonthNamesShort = {}));
