// @juicyllama/utils/dist/utils/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const Modules_1 = require("./Modules");
const Env_1 = require("./Env");
const DEFAULT_LOG_LEVEL = 3;
class Logger {
    data(key, value) {
        if (Env_1.Env.IsNotTest()) {
            let Bugsnag;
            if (Modules_1.Modules.bugsnag.isInstalled) {
                Bugsnag = Modules_1.Modules.bugsnag.load();
                Bugsnag.addMetadata(key, value);
            }
        }
        this.verbose(`[${key}]=>${JSON.stringify(value)}`);
    }
    error(message, ...optionalParams) {
        let colored = `\x1b[31m`;
        if (optionalParams) {
            message = this.processParams(message, optionalParams);
        }
        colored += `${message}\x1b[0m`;
        if (this.getLogLevel() <= 5) {
            console.error(colored);
        }
        if (Env_1.Env.IsNotTest()) {
            switch (message) {
                case 'Unexpected token o in JSON at position 1':
                case 'Unexpected token o in JSON at position 1SyntaxError: Unexpected token o in JSON at position 1':
                    break;
                default:
                    if (Modules_1.Modules.bugsnag.isInstalled) {
                        Modules_1.Modules.bugsnag.load().then(Bugsnag => {
                            Bugsnag.notify(new Error(message));
                        });
                    }
            }
        }
    }
    warn(message, ...optionalParams) {
        let colored = `\x1b[33m`;
        if (optionalParams) {
            message = this.processParams(message, optionalParams);
        }
        colored += `${message}\x1b[0m`;
        if (this.getLogLevel() <= 4) {
            console.warn(colored);
        }
    }
    log(message, ...optionalParams) {
        let colored = `\x1b[32m`;
        if (optionalParams) {
            message = this.processParams(message, optionalParams);
        }
        colored += `${message}\x1b[0m`;
        if (this.getLogLevel() <= 3) {
            console.log(colored);
        }
    }
    debug(message, ...optionalParams) {
        let colored = `\x1b[35m`;
        if (optionalParams) {
            message = this.processParams(message, optionalParams);
        }
        colored += `${message}\x1b[0m`;
        if (this.getLogLevel() <= 2) {
            console.debug(colored);
        }
    }
    verbose(message, ...optionalParams) {
        let colored = `\x1b[36m`;
        if (optionalParams) {
            message = this.processParams(message, optionalParams);
        }
        colored += `${message}\x1b[0m`;
        if (this.getLogLevel() <= 1) {
            console.debug(colored);
        }
    }
    status() {
        this.log(`--------- Logging Status ---------`);
        this.log(`LOG_LEVEL=${this.getLogLevel()}`);
        this.error(`This is an error`);
        this.warn(`This is a warning`);
        this.log(`This is a log`);
        this.debug(`This is a debug`);
        this.verbose(`This is a verbose`);
        this.log(`------- Logging Status End -------`);
    }
    table(data) {
        if (Env_1.Env.IsDev()) {
            console.table(data);
        }
    }
    getLogLevel() {
        return process.env.LOG_LEVEL ? Number(process.env.LOG_LEVEL) : DEFAULT_LOG_LEVEL;
    }
    processParams(message, optionalParams) {
        try {
            for (const param of optionalParams) {
                switch (typeof param) {
                    case 'undefined':
                        break;
                    case 'object':
                        try {
                            message += ` ${JSON.stringify(param)}`;
                        }
                        catch (e) {
                            message += ` **Logger Error** Failed to read JSON object`;
                        }
                        break;
                    case 'string':
                    case 'boolean':
                        message += param;
                        break;
                    default:
                        this.error(`typeof param of optionalParams ${typeof param} not handled`);
                }
            }
        }
        catch (e) {
            this.error(e.message);
            if (Modules_1.Modules.bugsnag.isInstalled) {
                Modules_1.Modules.bugsnag.load().then(Bugsnag => {
                    return Bugsnag.notify(new Error(e));
                });
            }
        }
        return message;
    }
}
exports.Logger = Logger;
