// @juicyllama/utils/dist/utils/Measurements.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Measurements = void 0;
class Measurements {
    static LbsToKgs(number) {
        return Number((Number(number) * 0.453592).toFixed(2));
    }
    static KgsToLbs(number) {
        return Number((Number(number) * 2.20462).toFixed(2));
    }
    static GramsToOz(number) {
        return Number((Number(number) * 0.035274).toFixed(2));
    }
    static FtInchesToCm(number) {
        const [feet, inches] = number.toString().split('.');
        const totalInches = Number(feet) * 12 + Number(inches);
        return Number((Number(totalInches) * 2.54).toFixed(2));
    }
    static CmToFtInches(number) {
        const totalInches = Number(number) / 2.54;
        const feet = Math.floor(totalInches / 12);
        const inches = Number(totalInches.toFixed(2)) - feet * 12;
        return Number(`${feet}.${inches.toFixed(0)}`);
    }
}
exports.Measurements = Measurements;
Measurements.Grams = {
    abbreviation: 'g',
    name: 'gram',
    name_plural: 'grams',
};
Measurements.Ounces = {
    abbreviation: 'oz',
    name: 'ounce',
    name_plural: 'ounces',
};
