// @juicyllama/utils/dist/utils/Modules.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Modules = void 0;
const Logger_1 = require("./Logger");
class Module {
    constructor(name) {
        this.name = name;
    }
    get isInstalled() {
        return Modules.isInstalled(this.name);
    }
    load() {
        return __awaiter(this, void 0, void 0, function* () {
            return Modules.load(this.name);
        });
    }
}
class Modules {
    static isInstalled(name) {
        var _a;
        if (Modules.isCurrentModule(name)) {
            return true;
        }
        try {
            const p = require.resolve(name, { paths: (_a = require.main) === null || _a === void 0 ? void 0 : _a.paths });
            return !!p;
        }
        catch (e) {
            const logger = new Logger_1.Logger();
            logger.debug(`[utils::modules::isInstalled] ${e.message}`, e);
            return false;
        }
    }
    static load(name) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            if (Modules.isCurrentModule(name)) {
                return require(process.cwd());
            }
            return (_a = require.main) === null || _a === void 0 ? void 0 : _a.require(name);
        });
    }
    static isCurrentModule(name) {
        return process.env.npm_package_name === name;
    }
}
exports.Modules = Modules;
Modules.apilayer = new Module('@juicyllama/app-apilayer');
Modules.aws = new Module('@juicyllama/app-aws');
Modules.everflow = new Module('@juicyllama/app-everflow');
Modules.mailchimp = new Module('@juicyllama/app-mailchimp');
Modules.mollie = new Module('@juicyllama/app-mollie');
Modules.openai = new Module('@juicyllama/app-openai');
Modules.pexels = new Module('@juicyllama/app-pexels');
Modules.scrapingbee = new Module('@juicyllama/app-scrapingbee');
Modules.shopify = new Module('@juicyllama/app-shopify');
Modules.shipbob = new Module('@juicyllama/app-shipbob');
Modules.slack = new Module('@juicyllama/app-slack');
Modules.wise = new Module('@juicyllama/app-wise');
Modules.wordpress = new Module('@juicyllama/app-wordpress');
Modules.xerocc = new Module('@juicyllama/app-xero-cc');
Modules.semrush = new Module('@juicyllama/app-semrush');
Modules.googleAnalytics = new Module('@juicyllama/app-google-analytics4');
Modules.datacache = new Module('@juicyllama/data-cache');
Modules.bugsnag = new Module('@bugsnag/js');
Modules.showdown = new Module('showdown');
